test-harness/infratests/validate.go

```diff
@@ -0,0 +1,128 @@
+// and will not fail if the *expected values* are a subset of the *actual values.*
+//
+// In the case that the *expected values* are found in the *actual values,* nil will be returned. Otherwise an
+// error describing the mismatch will be returned.
+//
+// Some examples (in pseudo-code):
+//	- Exact Match:
+//		a := {"key1":[{"key2" : "foo"}, {"key2" : "bar"}]}
+//		verifyTargetsExistInMap(a, a) --> nil
+//
+//	- Subset Match #1:
+//		a := {"key1":[{"key2" : "foo"}, {"key2" : "bar"}]}
+//		b := {"key1":[{"key2" : "foo"}]}
+//		verifyTargetsExistInMap(a, b) --> nil
+//
+//	- Subset Match #2:
+//		a := {"key1":[{"key2" : "foo"}, {"key2" : "bar"}]}
+//		b := {"key1":[]}
+//		verifyTargetsExistInMap(a, b) --> nil
+//
+//	- Subset Match #3:
+//		a := {"key1":{"key2": "foo", "key3", "bar"}}
+//		b := {"key1":{"key3", "bar"}}
+//		verifyTargetsExistInMap(a, b) --> nil
+//
+//	- Mismatch #1:
+//		a := {"key1":{"key2": "foo", "key3", "bar"}}
+//		b := {"key1":[]}
+//		verifyTargetsExistInMap(a, b) --> ERROR: wrong type for key `key1`
+//
+//	- Mismatch #2:
+//		a := {"key1":{"key2": "foo", "key3", "bar"}}
+//		b := {"key1":{"key4": "foo"}}
+//		verifyTargetsExistInMap(a, b) --> ERROR: `key4` not found
+//
+//	- Mismatch #3:
+//		a := {"key1":{"key2": "foo", "key3", "bar"}}
+//		b := {"key1":{"key2": "bar"}}
+//		verifyTargetsExistInMap(a, b) --> ERROR: wrong value for `key2`
+//
+// The algorithm used to do the equality check is to execute a BFS search in parallel for both maps. In the case
+// that a mismatch (the values do not match, or the shape of the maps is different in a way that indicates a mismatch)
+// is identified, the search will be terminated. If no mismatches are found then `nil` is returned
+func verifyTargetsExistInMap(dataSource map[string]interface{}, searchTargets map[string]interface{}) error {
+	for targetKey := range searchTargets {
+		candidateMatch := dataSource[targetKey]
+		target := searchTargets[targetKey]
+
+		// both maps should contain the target search key
+		if candidateMatch == nil || target == nil {
+			return fmt.Errorf("Unexpectedly could not find key '%s'", targetKey)
+		}
+
+		// the values for the key should be the same type
+		if !isSameType(candidateMatch, target) {
+			return fmt.Errorf("Unexpectedly found type '%T' instead of '%T'", candidateMatch, target)
+		}
+
+		// the key is found and both values are of the same type. time to look for a subset match
+		switch typedTarget := target.(type) {
+		case bool, float32, float64, int, string:
+			if typedTarget != candidateMatch {
+				return fmt.Errorf("Expected %s but got %s", typedTarget, candidateMatch)
+			}
+		case []interface{}:
+			if err := verifyTargetsExistInList(candidateMatch.([]interface{}), typedTarget); err != nil {
+				return err
+			}
+		case map[string]interface{}:
+			if err := verifyTargetsExistInMap(candidateMatch.(map[string]interface{}), typedTarget); err != nil {
```

Comment:
I don't think it's a worry here, but there is a recursion depth limit... https://golang.org/pkg/runtime/debug/#SetMaxStack

Reply:
hah, I'd love to see the TF plan where this becomes an issue. I think the max recursion depth of this algo will be ~ `max map nesting + max list length` so as you say, likely not a concern here.